_posts/2020-09-06-bench-rust-kotlin-microservices.md

```diff
@@ -0,0 +1,445 @@
+  (12 [high-density vCPUs](https://cloud.google.com/blog/products/compute/understanding-dynamic-resource-management-in-e2-vms), 6 GB memory):
+  - Runs a stock Dockerised Elasticsearch 7.8.0 instance, storage backend for the microservices.
+  - Maximum CPU utilisation during short peaks never exceeded 80%.[^scale]
+
+[^scale]: I have scaled the number of Elasticsearch CPU cores up until it ceased to be a bottleneck,
+    arriving at 12.
+
+## The Contenders
+
+1. `locations-kt-http4k`: [code on GitLab](https://gitlab.com/gooutopensource/locations-kt-http4k/-/tree/kt-http4k)
+   - lang: **Kotlin** v1.3.72
+   - framework: [**Http4k**][http4k] v3.163.0 with *Undertow* server engine v2.0.22[^http4k-perf]
+   - programming model: classic **synchronous** (threads & blocking I/O calls)
+
+2. `locations-kt-ktor`: [code on GitLab](https://gitlab.com/gooutopensource/locations-kt-ktor/-/tree/kt-ktor).
+   - lang: **Kotlin** v1.3.70 
+   - framework: [**Ktor**][ktor] v3.163.0 with *Netty* server engine v4.1.44
+   - programming model: **async** (Kotlin coroutines)
+
+3. `locations-rs`: [code on GitHub](https://github.com/strohel/locations-rs/tree/rs-actix).
+   - lang: **Rust** v1.45.2
+   - framework: [**Actix** web][actix] v2.0
+   - programming model: **async** (Rust *async/await* with [Tokio][tokio] runtime v0.2.22 through `actix-rt`)
+
+[tokio]: https://tokio.rs/
+
+[^http4k-perf]: A newer http4k version 3.258.0 with Undertow 2.1.3 very slightly regressed in
+    performance, see http4k server engine benchmarks.
+
+### The Story of `locations-kt-http4k`
+
+Http4k Locations service was written by [@**goodhoko**][goodhoko] at [GoOut][goout].
+It ended up becoming the production implementation at GoOut,
+and is thus most complete along with extras like Swagger serving.
+
+[goodhoko]: https://github.com/goodhoko
+
+Http4k [allows one to select from a number of server backends](https://www.http4k.org/guide/modules/servers/),
+so we first need to do a qualification benchmark and select the best performing engine to be fair.
+
+See [**Kttp4k server engine results on a dedicated page**](https://storage.googleapis.com/strohel-pub/bench-http4k-server-engines-pub/bench-results.html).
+Line labels on the graph correspond to [tags in the locations-kt-http4k repository](https://gitlab.com/gooutopensource/locations-kt-http4k/-/tags).
+
+One nice side-effect of this benchmarking was that
+I have [discovered a cause](https://github.com/http4k/http4k/issues/141#issuecomment-679330954)
+of the [**Netty** backend performance issue](https://github.com/http4k/http4k/issues/141),
+and [@**daviddenton** was quick to fix it](https://github.com/http4k/http4k/pull/480).
+The `netty-updfix-*` graphs, which exhibit some nice latency characteristics, already contain the fix.
+It was later released in http4k v3.259.0 (be sure to upgrade if you use this backend).
+
+**Apache HttpCore** version 4.4, labelled as `apache4-*`, was most CPU-efficient backend
```

Comment:
`apache4-*` apod. Ja jsem asi pitomej, ale nejak jsem nedokazal rozklíčovat ten placeholder `*`.

Reply:
Hmmm, to má pojednávat o těch labelech na https://storage.googleapis.com/strohel-pub/bench-http4k-server-engines-pub/bench-results.html - jakože `apche4-1`, `apache4-2`, ...

Ono je trochu slabý místo toho blog postu že tady rozebírám něco, co je vidět jen po kliknutí na to `See Kttp4k server engine results on a dedicated page`.

Možná tam šoupnout preview obrázek, aby to bylo jasnější?